agent/consul/stream/event_publisher.go

```diff
@@ -100,12 +100,41 @@ func NewEventPublisher(handlers SnapshotHandlers, snapCacheTTL time.Duration) *E
 		subscriptions: &subscriptions{
 			byToken: make(map[string]map[*SubscribeRequest]*Subscription),
 		},
-		snapshotHandlers: handlers,
+		snapshotHandlers: make(map[Topic]SnapshotFunc),
 	}
 
 	return e
 }
 
+// RegisterHandler will register a new snapshot handler function. If a handler is
+// already registered for this topic, then all open subscriptions for the topic
+// will be force closed and all snapshots related to the topic will be evicted
+// from the cache.
+func (e *EventPublisher) RegisterHandler(topic Topic, handler SnapshotFunc) error {
+	if topic.String() == "" {
+		return fmt.Errorf("the topic cannnot be empty")
+	}
+
+	if _, found := e.snapshotHandlers[topic]; found {
+		return fmt.Errorf("a handler is already registered for the topic: %s", topic.String())
+	}
+
+	e.snapshotHandlers[topic] = handler
+
+	return nil
+}
+
+func (e *EventPublisher) RefreshTopic(topic Topic) error {
+	if _, found := e.snapshotHandlers[topic]; !found {
+		return fmt.Errorf("topic %s is not registered", topic)
+	}
+
+	e.forceEvictByTopic(topic)
+	e.subscriptions.closeAllByTopic(topic)
+
+	return nil
+}
+
```

Comment:
Do we need to hold `e.lock` in these methods (particularly `RefreshTopic`). Or are they guarded by a higher-level lock?

Reply:
Not here but I did add one in `forceEvictByTopic`.

Access to the snapshotHandlers map isn't locked because the expectation is that all handlers get registered prior to executing Run (documented that in the RegisterHandler doc comment)

Reply:
Nice, thanks for clarifying!